### DISCUSSION TITLE
Electronic Receipts

### DISCUSSION THREAD

**Discussion by iarn-sl** [Unicontsoft/UcsFiscalPrinters#23] [Ideas]
Hi,
would you consider implementing electronic receipts.
For example Datecs FP-2000 can be set for e-shop and the command differs a bit, if you're willing to implement I can get the protocol and all. It can return a base64 encoded image of the fiscal receipt in the response i guess.

**Comment by wqweto** [comment]
> It can return a base64 encoded image of the fiscal receipt in the response i guess.

Electronic Receipts are currently supported by setting `DisablePrinting` to `true` in `POST /printers/:printer_id/receipt` request body (i.e. `cFiscalPrinter.PrintReceipt` method)

Some device models (some Tremol FPs for instance) can return base64 encoded QR barcode but you can render similar barcode using any QR library -- just follow the format of the string encoded using ReceiptNo, ReceiptDateTime, Fiscal Device Number and Fiscal Memory Number as returned by the POST request above.

Just let me know what base64 encoded content you are talking about -- whole receipt, barcode only or something else. Which command you are thinking about (might be already supported)?

cheers,
&lt;/wqw>

**Reply by iarn-sl** [reply]
I tried with FP-2000 device set for electronic receipts ( I think 21 ), but was unable to get it to work not to print out the receipt.
I also have a Tremol FP03, I'll try on it too and send results.
Would you, please, post an example request.

I was thinking the full receipt content as base64 as it can be send. Datecs have some application for electronic receipts, but I have no idea how it works and as long as I remember it is a DLL only.

p.s. If you need any tests on those device models and I can be of any help, I'll gladly be of service.

**Reply by wqweto** [reply]
Here is a sample **data-utf8.txt** for a non-printing sales receipt
```
{
    "ReceiptType": "Sale",
    "DisablePrinting ": true,
    "Operator": {
        "Code": "1",
        "Name": "Иван Петров",
        "Password": "****"
    },
    "Rows": [
        {
            "ItemName": "Продукт 1",
            "Price": 12.34,
        }
    ]
}
```
. . . which can be printed using **curl.exe** like this
```
c:> curl -X POST http://localhost:8192/printers/DT123456/receipt --data-binary @data-utf8.txt
```
. . . where **DT123456** is the serial number of the fiscal printer and **UcsFPHub.exe** is started with default settings (auto-detect devices and http listener on port 8192/tcp).

**Reply by iarn-sl** [reply]
It still prints out the receipt.
Tried on Hub v0.1.50-3838b16 and v0.1.49-67ba798

On both devices I get a printout

tried "DisablePrinting ": true and trimmed "DisablePrinting": true..

Response :
{
	"Ok": true,
	"ReceiptNo": "47",
	"ReceiptDateTime": "2022-04-18 11:45:37",
	"DeviceSerialNo": "DT123456",
	"FiscalMemoryNo": "02123456",
	"ReceiptAmount": 12.34
}

Device data:
{
    "DeviceSerialNo": "DT123456",
    "FiscalMemoryNo": "02123456",
    "DeviceProtocol": "DATECS",
    "DeviceModel": "FP-2000",
    "FirmwareVersion": "2.00BG 19FEB20 0920",
    "DeviceString": "Port=COM7;Speed=115200;Protocol=DATECS;DeviceSerialNo=DT123456;FiscalMemoryNo=02123456;ModelGroup=0",
    "CharsPerLine": 48,
    "CommentTextMaxLength": 46,
    "ItemNameMaxLength": 40,
    "TaxNo": "***",
    "TaxCaption": "ЕИК",
    "DeviceHost": "***",
    "DevicePort": "COM7",
    "Autodetected": true
}


TREMOL FP03-KL-V2 :

    "DeviceProtocol": "TREMOL",
    "DeviceModel": "TREMOL FP03-KL-V2",
    "FirmwareVersion": "К.С.B347",
    "DeviceString": "Port=COM4;Speed=19200;Protocol=TREMOL;DeviceSerialNo=ZK123456;FiscalMemoryNo=50123456;ModelGroup=2",

{
	"Ok": true,
	"ReceiptNo": "37",
	"ReceiptDateTime": "2022-04-18 10:54:00",
	"DeviceSerialNo": "*",
	"FiscalMemoryNo": "*",
	"ReceiptAmount": 12.34
}

DeviceSerialNo/FiscalMemoryNo modified for the post.

**Reply by wqweto** [reply]
> and trimmed "DisablePrinting": true..

Oops, sorry about that :-))

**Reply by iarn-sl** [reply]
:)

I tried all sorts of things and still can not get either one to not print a paper receipt and/or get a different response.

Can I somehow set a debug log and higher log verbosity in UscFPHub to see if InitDisablePrinting is being triggered and/or see the RAW command being sent?

I'll try to contact Tremol to see if the device needs firmware change to work for electronic receipts, but the FP-2000 is set for electronic receipts.

**Reply by wqweto** [reply]
I peeked at the implementation and it seems `InitDisablePrinting` is implemented only for DAISY devices on ISL protocol (the one used by legacy DATECS devices too) so unfortunately it's not documented in the protocol description how can an FP-2000 be used with electronic receipts.

For the new DATECS protocol (the one used by so called X devices like D-25X) the setting is implemented ok IMO.

For TREMOL devices the setting is implemented in the protocol but it looks like the device has to be allowed to disable printing. Not sure how this is configured and/or authorized by the manufacturer.

**Reply by wqweto** [reply]
This is the command for Electronic Receipt in TREMOL protocol
<span>             |<span>
---                |---
2.6.4. Command: | 30h / 0 – Open ELECTRONIC Fiscal sales receipt
input: | <OperNum[1..2]> <;> <OperPass[6]> <;> <ReceiptFormat[1]> <;> <PrintVAT[1]> <;> <FiscalRcpPrintType[‘8’]> {<’$’> <UniqueReceiptNumber[24]>}
output: | ACK
FPR operation: | Opens an postponed electronic fiscal receipt with 1 minute timeout assigned to the specified operator number and operator password, parameters for receipt format, print VAT, printing type and unique receipt number.
Input data:
OperNum (Operator Number) | Symbols from 1 to 20 corresponding to operator's number
OperPass (Operator Password) | 6 symbols for operator's password
ReceiptFormat (Format) | 1 symbol with value:<br>- '1' - Detailed<br> - '0' - Brief
PrintVAT (VAT included in the receipt) | 1 symbol with value:<br> - '1' – Yes<br> - '0' – No
**FiscalRcpPrintType** | **1 symbol with value '8' for Postponed printing of electronic receipt**
UniqueReceiptNumber | Up to 24 symbols for unique receipt number.<br> NRA format: XXXХХХХХ-ZZZZ-YYYYYYY where:<br> * ХХХХХХXX – 8 symbols [A-Z, a-z, 0-9] for individual device number,<br> * ZZZZ – 4 symbols [A-Z, a-z, 0-9] for code of the operator,<br> * YYYYYYY – 7 symbols [0-9] for next number of the receipt

It seems the implementation correctly sends `FiscalRcpPrintType` with value `8` but probably your FP ignores this flag if currently not configured to allow non-printing fiscalization.

**Reply by iarn-sl** [reply]
Thanks.
All I see in ver.200 proto for FP-200 is : 
 2Bh (43) SET THE HEADER, FOOTER AND PRINT OPTIONS
... 
‘^’ Set the type of voucher generation (0 - normal print document, 1- electronic document
without print). This option works only for devices registered as type 21 / E-shop /.
3Bh ( 59 ) - ЧЕТЕНЕ НА ДАННИ ЗА ЕЛЕКТРОНЕН ДОКУМЕНТ /САМО ЗА ТИП 21
 Data area: ["*"] | ["E"<DocNo>]
Response: "F" | "*" |"P"<Data>
Input data:
If "E": Finds a document with <DocNo> only if it is an electronic document.
If there is a "*" finds the last issued fiscal or reversal receipt before a daily report with clearing is made.

Does this looks like anything in the X / ISL ?
I have very limited experience and I'm not much of a .NET / VB guy, I'll try to find the protocol implementation and if I manage to somehow change it and get it working ( doubt it :) tho I'll try ) I'll send you the changes to clean the code and push it if you decide.

Thank you.

**Reply by wqweto** [reply]
IMO what looks more promising is command `85H (133) SERVICE-DISABLED PRINTING`

If this is allowed on type 21 devices to be executed without service jumper set (like they do now for the newest C protocol for X devices with command `45 (2Dh) Check for mode connection with PC`) then this would allow persisting fiscal receipts to fiscal memory without any output on paper.

I'll try to add a preliminary implementation to ISL protocol (for Datecs printers only) using this 85H (133) command.

**Reply by iarn-sl** [reply]
Thank you !
I'll be happy to assist with any testing on the devices i have available.
BTW Datecs have a sample app, and in a scriptlog.txt the following is available ( if you can make heads or tails of it ) :
```
//--- Информация от принтера ---
//Връща номер на принтера
255,IDnumber[\t][\t][\t]
//Връща последния използван номер на продажба(УНП)
255,UNP[\t][\t][\t] 
//Връша паролата на оператора.Въвежда се номер на оператор - 1.
255,OperPasw[\t]0[\t][\t]

//--- Продажба ---
//Проверява дали е отворена бележката
76
//Забранява печата на бележки от принтера
45,0[\t]
//Отваряне на фискална бележка
48,1[\t]0000[\t]DT123456-0000-0000001[\t]1[\t][\t]
//Добавя продукт към бележката
49,Продукт1[\t]1[\t]0.01[\t]1[\t][\t][\t]1[\t]
//Плащане на бележката
53,0[\t][\t]
//Затваря бележката
56
//Премахва забраната за печат.
45,1[\t]

//--- Генериране на електронна бележка ---
//Връща текстовата информация на QR кода, която се поставя на място [FISCAL QR CODE]
74,1[\t]
//Задаване на бележка за четене. Възможно е последната бележка да не се появи веднага в журнала, 
//хубаво е да се изпълни командата поне още 1 път.
125,0[\t]2774[\t]10[\t]
//Прочитане на 1 ред от бележката. Изпълнява се докато не се върне код на грешка -100003.
125,1[\t]

//--- Фактура ---
76
45,0[\t]
//Отваряне на фактура
48,1[\t]0000[\t]DT123456-0000-0000001[\t]1[\t]I[\t]
49,Продукт1[\t]1[\t]0.01[\t]1[\t][\t][\t]1[\t]
53,0[\t][\t]
//Данни за клиента. Последните 3 параметъра са задължителни
57,[\t][\t][\t][\t][\t]0[\t]1234567890123[\t]12345678901234[\t]
//Анулиране на бележка.
60
45,1[\t]
```
